src/adm/settings.rs

```diff
@@ -168,26 +177,46 @@ impl From<&mut Settings> for AdmSettings {
                     "partner_id", "adm_partner_id"
                 );
             } else {
-                panic!("Missing argument {}", "$new");
+                return Err(ConfigError::Message(format!("Missing argument {}", "$new")));
             }
         }
         if settings.adm_settings.is_empty() {
-            return Self::default();
+            return Ok(Self::default());
         }
-        if Path::new(&settings.adm_settings).exists() {
-            if let Ok(f) = File::open(&settings.adm_settings) {
-                return serde_json::from_reader(f).expect("Invalid ADM Settings file");
+        let mut settings_str = settings.adm_settings.clone();
+        if Path::new(&settings_str).exists() {
+            if let Ok(mut f) = File::open(&settings.adm_settings) {
+                settings_str = String::new();
+                f.read_to_string(&mut settings_str).map_err(|e| {
+                    ConfigError::Message(format!(
+                        "Could not read {}: {:?}",
+                        settings.adm_settings, e
+                    ))
+                })?;
             }
         }
         let adm_settings: AdmSettings =
-            serde_json::from_str(&settings.adm_settings).expect("Invalid ADM Settings JSON string");
+            serde_json::from_str(&settings_str).expect("Invalid ADM Settings JSON string");
         for (adv, filter_setting) in &adm_settings {
             if filter_setting
                 .include_regions
                 .iter()
                 .any(|region| region != &region.to_uppercase())
             {
-                panic!("Advertiser {:?} include_regions must be uppercase", adv);
+                return Err(ConfigError::Message(format!(
+                    "Advertiser {:?} include_regions must be uppercase",
+                    adv
+                )));
+            }
+            if filter_setting.advertiser_urls.iter().any(|filter| {
+                if let Some(ref paths) = filter.paths {
+                    return paths
+                        .iter()
+                        .any(|path| path.matching == "prefix" && !path.value.ends_with('/'));
```

Comment:
While you're refactoring the code here, could you also add a check for "exact" values? Specifically, valid "exact" values should start with '/'.

Reply:
To make things easier, I think we can just use a single error message to capture both, such as `Advertiser {:?} advertiser_urls contain invalid path values`